src/allreduce_robust.h

```diff
@@ -519,6 +579,12 @@ o   *  the input state must exactly one saved state(local state of current node)
   int result_buffer_round;
   // result buffer of all reduce
   ResultBuffer resbuf;
+  // current cached allreduce/braodcast sequence number
+  int cur_cache_seq;
+  // result buffer of cached all reduce
+  ResultBuffer cachebuf;
+  // key of each cache entry
+  ResultBuffer lookupbuf;
```

Comment:
so it essentially doubles even triples the memory footprint with in an iteration ?

in XGB, we save stats for each node in a tree, and now, we have to save it in rabit layer again?

Reply:
can you give an example PR on XGB side about how to leverage this feature?

Reply:
size of cache is bounded by number of keys user specificly defined to cache (those are not always hits in each iteration like bootstrap or DMatrix loading) 

The way rabit works without cache is it assume each iteration runs exactly same set of allreduce/broadcast operations. If one iteration complete with checkpoint succeed, it reset seq to 0. 

When a node failed, it loads latest checkpoint and reset seq no on that node to 0 which will try to connect to other nodes and align on seq no. As you might imagine, if one side is doing bootstrap and try allreduce with number of colums while others are trying to allreduce with histograms, it wouldn't work.

https://github.com/dmlc/xgboost/pull/4636/files#diff-bc273f37600ddbaef4874e160c1e7ca3R227

Reply:
In term of memory footprint overhead, approx method will cost 2 unit64_t bytes, fast histogram will be one hist_init size + 2 unit64_t + a float

Reply:
> In term of memory footprint overhead, approx method will cost 2 unit64_t bytes, fast histogram will be one hist_init size + 2 unit64_t + a float

this is not true according to https://github.com/dmlc/xgboost/pull/4636/files#diff-daec1d0ffe267847ee82f8b885a06dbbR179. 

every feature will have a sketch and similarly you also need to cache results of histogram corresponding to sketch (which I didn't find in your example PR)

Reply:
This is runtime log of fast histogram, you can see there are two d_colmn cache call from loading dmatrix and one from seed in colmnsampler, so histogram init only called once.  As long as we would be able to align seq counter in all nodes, we don't need to cache anything further.

`
[0]@@@Hit Mock Error:AllReduce
/home/cq/xgboost/xgboost machinehist.conf mock=0,10,1,0 DMLC_NUM_ATTEMPT=1
[21:13:45] start instance-xgboost:0
[21:13:45] Load part of data 0 of 10 parts
[21:13:45] 650x127 matrix with 14300 entries loaded from ../../demo/data/agaricus.txt.train
[0] RecoverExec caller GetCache seq 0
[21:13:45] Load part of data 0 of 10 parts
[21:13:45] 161x127 matrix with 3542 entries loaded from ../../demo/data/agaricus.txt.test
[0] RecoverExec caller GetCache seq 0
...
[0] RecoverExec caller LoadCheckPoint seq 0
...
[0] RecoverExec caller LoadCheckPoint seq 0
[0] load checkpoint global 4572 version 10
[0] RecoverExec caller LoadCheckPoint seq 0
...
[21:13:45] Tree method is specified to be 'hist' for distributed training.
[0] RecoverExec caller GetCache seq 0
[0] RecoverExec caller GetCache seq 0
...
[0] RecoverExec caller Allreduce seq 0
[1] RecoverExec caller Allreduce seq 1
...
[0] RecoverExec caller Allreduce seq 1
[9] RecoverExec caller Allreduce seq 1
[1] RecoverExec caller Allreduce seq 1
[8] RecoverExec caller Allreduce seq 1
[7] RecoverExec caller Allreduce seq 1
[2] RecoverExec caller Allreduce seq 1
[6] RecoverExec caller Allreduce seq 1
[3] RecoverExec caller Allreduce seq 1
[4] RecoverExec caller Allreduce seq 1
[5] RecoverExec caller Allreduce seq 1
...
`

Reply:
I agree it should be called for only once for caching sketch, but 

1. it still nearly doubles the memory footprint and `approx method will cost 2 unit64_t bytes` should not be true, as sketch is nearly the same in these two algorithms, 

2. I am also not convinced about why you didn't cache histogram itself in fast histogram and sketch in approx